engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/ConfluentAvroToJsonFormatterFactory.scala

```diff
@@ -0,0 +1,119 @@
+    implicit val valueEncoder: Encoder[V] = createValueEncoder(serializableRecord.consumerRecord.value, messageFormatter)
+    implicit val serializableRecordEncoder: Encoder[SerializableConsumerRecord[K, V]] = deriveEncoder
+    val consumerRecordEncoder: Encoder[AvroSerializableConsumerRecord[K, V]] = deriveEncoder
+
+    consumerRecordEncoder(serializableRecord).noSpaces.getBytes(StandardCharsets.UTF_8)
+  }
+
+  /**
+    * Step 1: Deserialize raw json bytes to AvroSerializableConsumerRecord[Json, Json] domain without interpreting key and value content.
+    * Step 2: Create key and value json-to-avro interpreter based on schema id's provided in json.
+    * Step 3: Use interpreter to create raw kafka ConsumerRecord
+    * @tparam K - key type passed from KafkaAvroSourceFactory, used to determine which datumReaderWriter use (e.g. specific or generic)
+    * @tparam V - value type passed from KafkaAvroSourceFactory, used to determine which datumReaderWriter use (e.g. specific or generic)
+    */
+  protected def doParseRecord[K: ClassTag, V: ClassTag](kafkaConfig: KafkaConfig,
+                                                        createReader: String => ConfluentAvroMessageReader)
+                                                       (topic: String, bytes: Array[Byte]): ConsumerRecord[Array[Byte], Array[Byte]] = {
+    implicit val serializableRecordDecoder: Decoder[SerializableConsumerRecord[Json, Json]] = deriveDecoder
+    val consumerRecordDecoder: Decoder[AvroSerializableConsumerRecord[Json, Json]] = deriveDecoder
+    val record = CirceUtil.decodeJsonUnsafe(bytes)(consumerRecordDecoder)
+
+    def serializeKeyValue(keyOpt: Option[Json], value: Json): (Array[Byte], Array[Byte]) = {
+      val reader = createReader(topic)
+      val keyBytes = if (kafkaConfig.useStringForKey) {
+        keyOpt match {
+          // we handle strings this way because we want to keep result value compact and JString is formatted in quotes
+          case Some(j) if j.isString => j.asString.get.getBytes(StandardCharsets.UTF_8)
+          case Some(j) => j.noSpaces.getBytes(StandardCharsets.UTF_8)
+          case None => Array.emptyByteArray
+        }
+      } else {
+        val keySchema = record.keySchemaId.map(id => reader.schemaById(id)).getOrElse(throw new IllegalArgumentException("Error reading schema: empty schema id"))
+        keyOpt.map(keyJson => reader.readJson[K](keyJson, Option(keySchema), ConfluentUtils.keySubject(topic))).getOrElse(Array.emptyByteArray)
+      }
+      val valueSchema = reader.schemaById(record.valueSchemaId)
+      val valueBytes = reader.readJson[V](value, Option(valueSchema), ConfluentUtils.valueSubject(topic))
+      (keyBytes, valueBytes)
+    }
+    record.consumerRecord.toKafkaConsumerRecord(topic, serializeKeyValue)
+  }
+
+  private def createKeyEncoder[K: ClassTag](key: K, useStringForKey: Boolean, messageFormatter: ConfluentAvroMessageFormatter): Encoder[K] = {
```

Comment:
1. Why key is passed here instead of using from argument in `apply`?  2. Can encoders be created once in formatter class (see comment above) and used many times (for each record)?

Reply:
1 & 2 Fixed.